### PR TITLE
Add stream conformance tests for standalone streams

### PR DESCRIPTION
**Pull request by stephentoub**
I previously added a stream test suite for streams that could be created in a connected fashion, such that writing to one stream yields data that can be read by another.  That enabled testing of a bunch of streams like NetworkStream, SslStream, BufferedStream, etc., even FileStream when wrapped around a pipe.

This PR augments the suite for standalone streams that aren't connected in that fashion.  This includes MemoryStream, FileStream for on-disk files, UnmanagedMemoryStream, MemoryMappedViewStream, etc.  As with the previous PR, even though it's adding a lot of coverage, total line count is reduced, because it's consolidating tests that were unique to an individual test suite.

I also added a couple of tests to the existing connected suite.

cc: @geoffkizer, @carlossanlop, @Jozkee

### CODE REVIEW

File: src/libraries/Common/tests/Tests/System/IO/StreamConformanceTests.cs [hunk 1]
```diff
@@ -158,8 +340,8 @@ protected async Task ValidateMisuseExceptionsAsync(Stream stream)
                 Assert.Throws(UnsupportedReadWriteExceptionType, () => { stream.ReadByte(); });
                 Assert.Throws(UnsupportedReadWriteExceptionType, () => { stream.Read(new Span<byte>(new byte[1])); });
                 Assert.Throws(UnsupportedReadWriteExceptionType, () => { stream.Read(new byte[1], 0, 1); });
-                Assert.Throws(UnsupportedReadWriteExceptionType, () => { stream.ReadAsync(new byte[1], 0, 1); });
-                Assert.Throws(UnsupportedReadWriteExceptionType, () => { stream.ReadAsync(new Memory<byte>(new byte[1])); });
+                await Assert.ThrowsAsync(UnsupportedReadWriteExceptionType, () => stream.ReadAsync(new byte[1], 0, 1));
+                await Assert.ThrowsAsync(UnsupportedReadWriteExceptionType, async () => await stream.ReadAsync(new Memory<byte>(new byte[1])));
```
Comment (geoffkizer) [inline src/libraries/Common/tests/Tests/System/IO/StreamConformanceTests.cs:344]: Why are these two different?

Comment (geoffkizer) [reply src/libraries/Common/tests/Tests/System/IO/StreamConformanceTests.cs:344]: (similar below for Write)

Comment (geoffkizer) [reply src/libraries/Common/tests/Tests/System/IO/StreamConformanceTests.cs:344]: Ah is this because of ValueTask? Ok then.

Comment (stephentoub) [reply src/libraries/Common/tests/Tests/System/IO/StreamConformanceTests.cs:344]: ThrowsAsync expects a `Func<Task>`.  The second overload returns a `ValueTask`.  So it's either using async/await or add .AsTask().  I could make the first one match with async/await, as well, just wasn't necessary.

Comment (stephentoub) [reply src/libraries/Common/tests/Tests/System/IO/StreamConformanceTests.cs:344]: Right.

File: src/libraries/Common/tests/Tests/System/IO/StreamConformanceTests.cs [hunk 2]
```diff
@@ -314,173 +499,1016 @@ void ValidateDisposedException(Exception e, bool successAllowed = false)
...
+                Assert.Equal(cts.Token, oce.CancellationToken);
+            }
+
+            if (stream.CanWrite)
+            {
+                oce = await Assert.ThrowsAnyAsync<OperationCanceledException>(() => stream.WriteAsync(new byte[1], 0, 1, cts.Token));
+                Assert.Equal(cts.Token, oce.CancellationToken);
+
+                oce = await Assert.ThrowsAnyAsync<OperationCanceledException>(async () => { await stream.WriteAsync(new ReadOnlyMemory<byte>(new byte[1]), cts.Token); });
+                Assert.Equal(cts.Token, oce.CancellationToken);
```
Comment (geoffkizer) [inline src/libraries/Common/tests/Tests/System/IO/StreamConformanceTests.cs:524]: Just a thought: Maybe we should have some sort of Assert for cancellation specifically, that also checks the cancellationToken? Seems like this is a pretty common pattern here.

Comment (stephentoub) [reply src/libraries/Common/tests/Tests/System/IO/StreamConformanceTests.cs]: Added a helper.

**Pull request merged**